test/src/unit-cppapi-var-offsets.cc

```diff
@@ -787,17 +847,33 @@ TEST_CASE(
         CHECK_NOTHROW(query_w.submit());
         array_w.close();
 
-        // Assume no size for the extra element
-        std::vector<int32_t> attr_val(data.size());
-        std::vector<uint64_t> attr_off(element_offsets.size() - 1);
-
         // Submit read query
         Array array_r(ctx, array_name, TILEDB_READ);
         Query query_r(ctx, array_r, TILEDB_READ);
+
+        // Assume no size for the extra element
+        std::vector<int32_t> attr_val(data.size());
+        std::vector<uint64_t> attr_off(element_offsets.size() - 1);
         query_r.set_buffer("attr", attr_off, attr_val);
         query_r.set_subarray<int64_t>({1, 2, 1, 2});
-        // The query should fail
-        CHECK_THROWS(query_r.submit());
+
+        // First partial read because offsets don't fit
+        CHECK_NOTHROW(query_r.submit());
+        CHECK(query_r.query_status() == Query::Status::INCOMPLETE);
+        std::vector<int32_t> data_exp1 = {1, 2, 3, 0, 0, 0};
+        std::vector<uint64_t> data_off_exp1 = {0, 1, 3, 0};
+        CHECK(attr_val == data_exp1);
+        CHECK(attr_off == data_off_exp1);
+
+        // Second partial read
+        reset_read_buffers(attr_val, attr_off);
+        CHECK_NOTHROW(query_r.submit());
+        CHECK(query_r.query_status() == Query::Status::COMPLETE);
+        std::vector<int32_t> data_exp2 = {4, 5, 6, 0, 0, 0};
+        std::vector<uint64_t> data_off_exp2 = {0, 2, 3, 0};
+        CHECK(attr_val == data_exp2);
+        CHECK(attr_off == data_off_exp2);
```

Comment:
This looks correct to me, but it took me a minute to figure out because the C++ does not mutate the size of the `attr_val` or `attr_off`, which is why it has to pad with trailing zeroes.

To make this a little more obvious, could you `CHECK()` the expected number of returned elements with `query.result_buffer_elements["attr]`? For example:
https://github.com/TileDB-Inc/TileDB/blob/dev/examples/cpp_api/reading_incomplete.cc#L174

Reply:
Thanks for the hint, I agree it improves testing and it also revealed a small bug I just fixed :-)